# Reading Cloud-Optimized GeoTIFFs the Hard Way

In this notebook we will explore how one can read Cloud-Optimized GeoTIFFs (COGs) the hard way, i.e., by requesting and parsing byte ranges by hand. We'll query 

In [350]:
from __future__ import annotations

import dataclasses
import enum
import itertools
import struct
import urllib.request

from glob import glob
from pathlib import Path
from pprint import pprint
from typing import Any, Iterator, Literal, Self

import folium
import numpy as np
import shapely

from pystac_client import Client

In [230]:
# This is a mapping of the TIFF data types to the struct package's format charaters
# see https://docs.python.org/3/library/struct.html#format-characters

DATA_TYPES = {
    1: '1B',  # BYTE (uint8)
    2: '1s',  # ASCII (char[1])
    3: '1H',  # SHORT (uint16)
    4: '1I',  # LONG (uint32)
    5: '2I',  # RATIONAL (uint32[2])
    6: '1b',  # SBYTE (int8)
    7: '1B',  # UNDEFINED (uint8)
    8: '1h',  # SSHORT (int16)
    9: '1i',  # SLONG (int32)
    10: '2i',  # SRATIONAL (int32[2])
    11: '1f',  # FLOAT (float32)
    12: '1d',  # DOUBLE (float64)
    13: '1I',  # SUBIFD (uint32)
    # 14: '',
    # 15: '',
    16: '1Q',  # ? (uint64)
    17: '1q',  # ? (int64)
    18: '1Q',  # ? (uint64)
}

In [341]:
ENDIANNESS = {
    b'MM': '>',  # big endian
    b'II': '<',  # little endian
}

In [154]:
def get_tag_size(magic_number: int) -> Literal[12, 20]:
    match magic_number:
        case 42:  # TIFF
            return 12
        case 43:  # BigTIFF
            return 20
        case _:
            raise TypeError(f"Unsupported file type: magic number {magic_number}")  


def url_read_bytes(url: str, start: int, end: int) -> bytes:
    request = urllib.request.Request(
        url,
        headers={'Range': f'bytes={start}-{end-1}'},
    )
    with urllib.request.urlopen(request) as response:
        return response.read()

In [251]:
# Point of Interest
poi = shapely.Point(-121.695833, 45.373611)

In [252]:
# use folium to display vectors
# Several folium basemap tiles are available:
#   - OpenStreetMap
#   - Stamen Terrain
#   - Stamen Toner
#   - Stamen Watercolor
#   - CartoDB positron
#   - CartoDB dark_matter

location = (poi.coords[0][1], poi.coords[0][0])

map = folium.Map(
    location=location,
    tiles="Cartodb Positron",
)

folium.CircleMarker(
    location=location,
    fill=True,
    fill_opacity=0.6,
).add_to(map)

map

In [283]:
client = Client.open("https://earth-search.aws.element84.com/v1")

search = client.search(
    max_items=1,
    collections=['sentinel-2-c1-l2a'],
    intersects=poi,
    datetime='2023/2023',
    query=['eo:cloud_cover<10'],
    sortby=[{"direction": "desc", "field": "properties.datetime"}],
)
item = next(search.items())
pprint(item.to_dict())

{'assets': {'aot': {'file:checksum': '1220acf5f2c6389f16bdd5ccc980ece525d8bf58a7bc4cc94cd94db2208b5b064e07',
                    'file:size': 2163428,
                    'gsd': 20,
                    'href': 'https://e84-earth-search-sentinel-data.s3.us-west-2.amazonaws.com/sentinel-2-c1-l2a/10/T/FR/2023/12/S2B_T10TFR_20231223T190950_L2A/AOT.tif',
                    'proj:shape': [5490, 5490],
                    'proj:transform': [20, 0, 600000, 0, -20, 5100000],
                    'raster:bands': [{'data_type': 'uint16',
                                      'nodata': 0,
                                      'offset': 0,
                                      'scale': 0.001,
                                      'spatial_resolution': 20}],
                    'roles': ['data'],
                    'title': 'Aerosol optical thickness (AOT)',
                    'type': 'image/tiff; application=geotiff; '
                            'profile=cloud-optimized'},
            'blue': {'

In [256]:
folium.GeoJson(item).add_to(map)
map

In [257]:
href = item.assets['red'].href
print(href)

https://e84-earth-search-sentinel-data.s3.us-west-2.amazonaws.com/sentinel-2-c1-l2a/10/T/FR/2023/12/S2B_T10TFR_20231223T190950_L2A/B04.tif


In [260]:
first_bytes = url_read_bytes(href, 0, 8)
print(first_bytes)

b'II*\x00\xc0\x00\x00\x00'


In [261]:
# COG spec: https://github.com/cogeotiff/cog-spec/blob/master/spec.md

hdr = first_bytes[0:4]

TIFF uses the first two bytes of the file to encode the endianness of the file. This enables writers to use the most efficient endianess for their host system, if desired. In other words, readers must support reading big or little endian files, where writers can pick one endianness.

Big endian is encoded as `MM` (from Motorola processors), and little endian is encoded as `II` (from Intel processors).

In [262]:
# We can see that these markers are the same when read as big or little endian,
# because that just affects the order of the bytes in the two-byte words,
# not the order of the bits within the bytes.

# big endian signature
print(bin(struct.unpack('>H', b'MM')[0]))
print(bin(struct.unpack('<H', b'MM')[0]))

# little endian signature
print(bin(struct.unpack('>H', b'II')[0]))
print(bin(struct.unpack('<H', b'II')[0]))

0b100110101001101
0b100110101001101
0b100100101001001
0b100100101001001


In [263]:
# our image signature
print(first_bytes[0:2])
endianness = ENDIANNESS[first_bytes[0:2]]  # We'll need this later, so let's save it into a var now

b'II'


In [264]:
magic_number = struct.unpack(f'{endianness}H', first_bytes[2:4])[0]
magic_number

42

In [265]:
import sys
sys.byteorder

'little'

In [267]:
first_bytes[0:4].hex()

'49492a00'

In [268]:
chr(0x49)

'I'

In [269]:
print(0x002a)  # TIFF
print(0x002b)  # BigTIFF

42
43


In [270]:
ifd_offset_bytes = first_bytes[4:8]
print(ifd_offset_bytes)

b'\xc0\x00\x00\x00'


In [271]:
ifd_offset = struct.unpack(f'{endianness}I', ifd_offset_bytes)[0]
print(ifd_offset)

192


## Parsing the IFD

The IFD is a data structure composed of entries called tags (hence the name "Tag Image File Format"). The IFD doesn't start with the first tag entry, however. It begins with a 2-byte `unit16` value indicating the number of tags within the IFD. This value enables us, along with the IFD offset within the file, to read the entire sequence of tag bytes via `file_bytes[ifd_offset + 2:ifd_offset + (tags_count * tag_size)]`.

### Tag structure

In a standard TIFF, tags are a 12-byte sequence (so `tag_size` above is 12 bytes) of the following structure:

| Tag Bytes | Tag field name  | Field data type |
| --------- | --------------- | --------------- |
| 0 - 1     | `code`          | `uint16`        |
| 2 - 3     | `data_type`     | `uint16`        |
| 4 - 8     | `count`         | `uint32`        |
| 8 - 12    | `value`         | `char[4]`       |

In the case of BigTIFF files, each tag is a 20-byte sequence where the `count` and `value` are of type `uint64`.

The tag `code` field gives us a way to find the meaning of the tag `value`, as the `code` is an integer that maps to the tag name. The Library of Congress has [a handy table](https://www.loc.gov/preservation/digital/formats/content/tiff_tags.shtml) we can use to look up the tags by their codes.

#### Tag data types

The tag `data_type` is also an integer value, in this case mapping to the data type we can use to interpret `value` per the following table:

| `data_type` | Type Name | Data type   |
| ----------- | --------- | ----------- |
| 1           | BYTE      | `uint8`     |
| 2           | ASCII     | `char[1]`   |
| 3           | SHORT     | `uint16`    |
| 4           | LONG      | `uint32`    |
| 5           | RATIONAL  | `uint32[2]` |
| 6           | SBYTE     | `int8`      |
| 7           | UNDEFINED | `uint8`     |
| 8           | SSHORT    | `int16`     |
| 9           | SLONG     | `int32`     |
| 10          | SRATIONAL | `int32[2]`  |
| 11          | FLOAT     | `float32`   |
| 12          | DOUBLE    | `float64`   |
| 13          | SUBIFD    | `uint32`    |
| 14          | n/a       | n/a         |
| 15          | n/a       | n/a         |
| 16          | ?         | `uint64`    |
| 17          | ?         | `int64`     |
| 18          | ?         | `uint64`    |

(I believe data types 16, 17, and 18 are specifc to BigTIFF, but I have so far been unable to find confirmation either way.)

The `count` field tells us how many of the listed `data_type` make up the `value` of the tag. Note that even a `count` of just one for a `data_type` of, say 5, or two `uint32`s would not fit in a `value` in a standard TIFF file as `value` itself is only four bytes long. Similarly, a `count` greater than 4 with a `data_type` of 1 (`uint8`) would also be larger than can fit in `value`.

In such cases where `count * len_in_bytes(data_type) > 4`, `value` itself is not actually the tag value but an offset to the actual value within the file. The length of that value is given by the previous expression `count * len_in_bytes(data_type)`. Thus, to get the actual value we can read `file_bytes[value:value + (count * len_in_bytes(data_type))]`.

The IFD doesn't end with the last tag either. Each IFD contains a 4-byte (`uint32`) offset to the next IFD in the file (or 8-byte `uint64` in the case of BigTIFF). In the event an IFD is the last one in the file it will have a value of 0 for its next IFD offset. As a result, it should be possible to build a map of the complete contents of a TIFF by iterating through its IFDs and parsing their tags into some appropriate hierarchical data structure (TIFF --< IFDs --< Image segments) .

### Finding the tag count and reading the tag bytes

As mentioned, an IFD starts with a 2-byte `uint16` value indicating its number of tags. If we have an IFD's offset (`ifd_offset`) within the file--which for the first IFD we know is given to us as the first bytes in the file immediately following the TIFF header--then we also know that IFS's tag offset (`tags_start`) is given by `ifd_offset + 2`.

Parsing the tag count (`tags_count`) should simply be a matter of using `struct.unpack` to unpack the two tag count bytes into an integer (struct format char `H` for `uint16`). We need to make sure we use the endianness indicicated in the file header. `<` is little endian in `struct.unpack`, where `>` is big endian. Looking back, the proper endian character should have been saved into the `endianness` var for us back when we were inspecting the header bytes.

In [272]:
tags_start = ifd_offset + 2
tags_count = struct.unpack(f'{endianness}H', url_read_bytes(href, ifd_offset, tags_start))[0]
print(tags_count)

19


If we know the tag count and the tag size (12 bytes for TIFF, 20 for BigTIFF), then we can find the total number of bytes in the IFD's tags by `tag_count * tag_size`. From this we should be able to find the last byte of the tags with `tags_end = tags_start + (tag_count * tag_size)`, allowing us to read the tag bytes (`tags_bytes`) from the file.

It's also important to note that we can use the `tags_end` to know the offset of the next IFD offset, which a 4-byte value we can unpack into a `uint32` (for a standard TIFF).

In [289]:
# We can get the tag size for our file type using the `get_tag_size` function
tag_size = get_tag_size(magic_number)
tags_end = tags_start + (tags_count * tag_size)
tags_bytes = url_read_bytes(href, tags_start, tags_end)
next_ifd_offset = struct.unpack(f'{endianness}I', url_read_bytes(href, tags_end, tags_end + 4))[0]

print(tags_bytes)
print(next_ifd_offset)

b'\x00\x01\x03\x00\x01\x00\x00\x00\xe4*\x00\x00\x01\x01\x03\x00\x01\x00\x00\x00\xe4*\x00\x00\x02\x01\x03\x00\x01\x00\x00\x00\x10\x00\x00\x00\x03\x01\x03\x00\x01\x00\x00\x00\x08\x00\x00\x00\x06\x01\x03\x00\x01\x00\x00\x00\x01\x00\x00\x00\x15\x01\x03\x00\x01\x00\x00\x00\x01\x00\x00\x00\x1c\x01\x03\x00\x01\x00\x00\x00\x01\x00\x00\x00=\x01\x03\x00\x01\x00\x00\x00\x02\x00\x00\x00B\x01\x03\x00\x01\x00\x00\x00\x00\x04\x00\x00C\x01\x03\x00\x01\x00\x00\x00\x00\x04\x00\x00D\x01\x04\x00y\x00\x00\x00X\x07\x00\x00E\x01\x04\x00y\x00\x00\x00<\t\x00\x00S\x01\x03\x00\x01\x00\x00\x00\x01\x00\x00\x00\x0e\x83\x0c\x00\x03\x00\x00\x00\xca\x03\x00\x00\x82\x84\x0c\x00\x06\x00\x00\x00\xe2\x03\x00\x00\xaf\x87\x03\x00 \x00\x00\x00\x12\x04\x00\x00\xb1\x87\x02\x00\x1e\x00\x00\x00R\x04\x00\x00\x80\xa4\x02\x00 \x02\x00\x00\xaa\x01\x00\x00\x81\xa4\x02\x00\x02\x00\x00\x000\x00\x00\x00'
1136


### Parsing each tag

To parse each tag we need to find a way to split each tag's bytes out of the of the larger bytes string. Python gives us many valid ways of doing this. Let's start by using a list comprehension to split the tags bytes into a list of byte strings for each tag an see what those look like.

In [274]:
tag_bytes_list = [tags_data[i*tag_size:(i*tag_size)+tag_size] for i in range(len(tags_bytes)//tag_size)]
for tag_bytes in tag_bytes_list:
    print(tag_bytes)

b'\x00\x01\x03\x00\x01\x00\x00\x00\xe4*\x00\x00'
b'\x01\x01\x03\x00\x01\x00\x00\x00\xe4*\x00\x00'
b'\x02\x01\x03\x00\x01\x00\x00\x00\x10\x00\x00\x00'
b'\x03\x01\x03\x00\x01\x00\x00\x00\x08\x00\x00\x00'
b'\x06\x01\x03\x00\x01\x00\x00\x00\x01\x00\x00\x00'
b'\x15\x01\x03\x00\x01\x00\x00\x00\x01\x00\x00\x00'
b'\x1c\x01\x03\x00\x01\x00\x00\x00\x01\x00\x00\x00'
b'=\x01\x03\x00\x01\x00\x00\x00\x02\x00\x00\x00'
b'B\x01\x03\x00\x01\x00\x00\x00\x00\x04\x00\x00'
b'C\x01\x03\x00\x01\x00\x00\x00\x00\x04\x00\x00'
b'D\x01\x04\x00y\x00\x00\x00\xd6\x02\x00\x00'
b'E\x01\x04\x00y\x00\x00\x00\xf2\x00\x00\x00'
b'S\x01\x03\x00\x01\x00\x00\x00\x01\x00\x00\x00'
b'\x0e\x83\x0c\x00\x03\x00\x00\x00\x0c\x05\x00\x00'
b'\x82\x84\x0c\x00\x06\x00\x00\x00$\x05\x00\x00'
b'\xaf\x87\x03\x00 \x00\x00\x00T\x05\x00\x00'
b'\xb1\x87\x02\x00\x1e\x00\x00\x00\x94\x05\x00\x00'
b'\x80\xa4\x02\x00Q\x00\x00\x00\xba\x04\x00\x00'
b'\x81\xa4\x02\x00\x02\x00\x00\x000\x00\x00\x00'


#### Unpacking the tag values

With a way to extract each tag's bytes, we next need to use `struct.unpack` to extract the byte values into some we can use in python for the tag's `code`, `data_type`, `count`, and `value`. Remember that `code` and `data_type` are `uint16` values, which map to the struct `H` format. Look up the proper struct format values for `count` and `value` knowing what you know about the data types of those tag fields and verify if the format passed into `struct.unpack` in the example here is correct (feel free to consult the `DATA_TYPES` dict above or the struct docs directly).

For variety, this example implementation uses a `while` loop to extract the tag bytes. Each tag's fields are added into a dictionary indexed by the tag `code` to facilitate easy access in later code.

In [275]:
tags = {}
tag_index = 0

while tag_index < tag_count:
    try:
        tag_bytes = tags_bytes[tag_size * tag_index:(tag_size * (tag_index + 1))]
        tag_index += 1
    except IndexError:
        break

    code, data_type, count, value = struct.unpack(f'{endianness}HHI4s', tag_bytes)
    tags[code] = {
        'data_type': data_type,
        'count': count,
        'value': value,
    }

tags

{256: {'data_type': 3, 'count': 1, 'value': b'\xe4*\x00\x00'},
 257: {'data_type': 3, 'count': 1, 'value': b'\xe4*\x00\x00'},
 258: {'data_type': 3, 'count': 1, 'value': b'\x10\x00\x00\x00'},
 259: {'data_type': 3, 'count': 1, 'value': b'\x08\x00\x00\x00'},
 262: {'data_type': 3, 'count': 1, 'value': b'\x01\x00\x00\x00'},
 277: {'data_type': 3, 'count': 1, 'value': b'\x01\x00\x00\x00'},
 284: {'data_type': 3, 'count': 1, 'value': b'\x01\x00\x00\x00'},
 317: {'data_type': 3, 'count': 1, 'value': b'\x02\x00\x00\x00'},
 322: {'data_type': 3, 'count': 1, 'value': b'\x00\x04\x00\x00'},
 323: {'data_type': 3, 'count': 1, 'value': b'\x00\x04\x00\x00'},
 324: {'data_type': 4, 'count': 121, 'value': b'X\x07\x00\x00'},
 325: {'data_type': 4, 'count': 121, 'value': b'<\t\x00\x00'},
 339: {'data_type': 3, 'count': 1, 'value': b'\x01\x00\x00\x00'},
 33550: {'data_type': 12, 'count': 3, 'value': b'\xca\x03\x00\x00'},
 33922: {'data_type': 12, 'count': 6, 'value': b'\xe2\x03\x00\x00'},
 34735: {'data

#### Understanding tag codes

Now that we have TIFF tag values to look at, it would be good to mention the [Libray of Congress' guide to TIFF Tags](https://www.loc.gov/preservation/digital/formats/content/tiff_tags.shtml) again. We can use that lookup table to interpret each of the integer codes in a meaningful way. Note that some codes we will see in every file, while others may be specific to the way a file was encoded or the type of data it contains. Further, a number of the tags are specific to the GeoTIFF format and are required for such files, while some are used for metadata by GDAL and can generally be expected in a GeoTIFF (though not always of course).

For example, we should always expect to see 256, 257, 258, and 259:

| Code | Tag Name      | Tag Description              |
| ---- | ------------- | ---------------------------- |
| 256  | ImageWidth    | Number of image columns      |
| 257  | ImageLength   | Number of image rows         |
| 258  | BitsPerSample | Number of bits in each pixel |
| 259  | Compression   | Integer mapping to compression algorithm used for each image segment |

In the spirit of needing to consult various external lookup tables, the [Wikipedia entry for TIFF has a great table of possible compression formats and their integer values](https://en.wikipedia.org/wiki/TIFF#TIFF_Compression_Tag).

In [137]:
# TODO: don't parse them all. Show examples.
# Convert the tags dicts to a dict indexed on code.
# Match on certain expected codes to show how to parse.
# Leave it an exercise for the students to map each code to its name and figure out how to use it.

#### Unpacking the tag values

Recalling the earlier explanation about tag data types, counts, and values, we know that unpacking the tag values will not be the same for each tag given the differences in those three aforementioned tag fields across each of our different tags. For some tags that have a single count of a shorter data type we can unpack the tag `value` directly. But for longer values we'll have to use the tag `value` as an offset into the file to read the actual bytes to unpack.

We'll start with one of these easier examples and unpack the image size tags 256 and 257. Check the data types for these tags. What are the struct format chars for each? Will we need to unpack all four bytes of the `value` for either of these tags?

In [276]:
# image column count (width)
cols = struct.unpack('<H', tags[256]['value'][0:struct.calcsize('H')])[0]

# image row count (height)
rows = struct.unpack('<H', tags[257]['value'][0:struct.calcsize('H')])[0]

print(f'Image size is {cols} x {rows}')

Image size is 10980 x 10980


In the cases where the tag `value`'s four bytes are not sufficient to contain the whole tag value, parsing is a bit more complex. We not only need to find the struct format character (`struct_dtype`) and size for the tag's data type, but then we need to:

* use the data type size and the tag `count` to calculate how many bytes we need to read (`size`)
* unpack the `value` to get the actual value's byte offset in the file (`offset`)
* combine `size` and `offset` to get the byte range and read that out fo the file (giving us `values`)
* build the struct format string (`endianness + (struct_dtype * count)`) then unpack `values`

We'll preview this here with an example unpacking the tile offsets tag (324). The values we get out of this (`tile_offsets`) are the byte offsets for each image segment (tile) in the image represented by this IFD. We will be able to use these offsets in the next section to read the specific tile containing our POI (though we'll have unpack the rest of our tags and do a bit of math to figure out which one and what to do with the bytes).

In [282]:
tag = tags[324]
struct_dtype = DATA_TYPES[tag['data_type']]
size = tag['count'] * struct.calcsize(struct_dtype)
offset = struct.unpack(f'{endianness}I', tag['value'])[0]
values = url_read_bytes(href, offset, offset+size)
tile_offsets = struct.unpack(endianness + (struct_dtype * tag['count']), values)

for idx, tile_offset in enumerate(tile_offsets):
    print(f"Offset tile {idx}: {tile_offset}")

Offset tile 0: 55962680
Offset tile 1: 57411167
Offset tile 2: 58810332
Offset tile 3: 60222446
Offset tile 4: 61651003
Offset tile 5: 63054996
Offset tile 6: 64463518
Offset tile 7: 66025043
Offset tile 8: 67523672
Offset tile 9: 68987825
Offset tile 10: 70439668
Offset tile 11: 71480485
Offset tile 12: 72831139
Offset tile 13: 74191906
Offset tile 14: 75556803
Offset tile 15: 76922917
Offset tile 16: 78346396
Offset tile 17: 79767466
Offset tile 18: 81177106
Offset tile 19: 82626646
Offset tile 20: 84045343
Offset tile 21: 85436959
Offset tile 22: 86443457
Offset tile 23: 87744763
Offset tile 24: 89128625
Offset tile 25: 90516041
Offset tile 26: 91896145
Offset tile 27: 93323921
Offset tile 28: 94699513
Offset tile 29: 96054131
Offset tile 30: 97398784
Offset tile 31: 98768288
Offset tile 32: 100117176
Offset tile 33: 101099165
Offset tile 34: 102475360
Offset tile 35: 103914015
Offset tile 36: 105337327
Offset tile 37: 106767167
Offset tile 38: 108155055
Offset tile 39: 109496290
Of

### Questions

* Refer back to the STAC item and see if the `file` STAC extension is in use. Is the file size listed for the COG asset your are examining, and if so how close to the end of the file do these tiles appear to get?
* Can you use the unpacking examples to create a generalized approach to unpacking the tag values and apply that to the rest of the tags in the IFD? The next section will have you unpack all the tags, so finding a quick an efficient way to do this might be helpful.

### Answers

* The COG is 218,693,282 bytes. The last tile starts at offset 217,929,856. The difference between those two is 763,426 bytes. Looking at the offset of the second-to-last tile, 216,879,023, we can see that tile to be 217,929,856 - 216,879,023 = 1,050,833 bytes in size. So 763,426 bytes is will within the expected size of a tile, and because of that we can reasonably conclude that this tile is the last data in the file.

  Note this aligns with our understanding of the structure of a COG: the IFDs are in the beginning of the file, and the actual image data follows with the full resolution data at the end (each overview progressively lower resolution stacked on top from highest resolution at the end to lowest at the top).
* We can modify the more complex tag unpack case as a starting point to implement a function that handles unpacking all types of tag values. It could be great if we use it to parse out the tag from the raw bytes into a complete object, even. We could go so far as to start with the TIFF itself and build up a data structure that parses all the IFDs out of the file for us. Let's see what that could look like:

In [393]:
class Endianness(bytes, enum.Enum):
    BIG_ENDIAN = b'MM'
    LITTLE_ENDIAN = b'II'

    @property
    def unpack_char(self: Self) -> str:
        match self:
            case Endianness.BIG_ENDIAN:
                return '>'
            case Endianness.LITTLE_ENDIAN:
                return '<'


@dataclasses.dataclass
class TIFFBytes:
    data: bytes
    endianness: Endianness

    def unpack(self: Self, format: str) -> tuple[Any, ...]:
        return struct.unpack(f'{self.endianness.unpack_char}{format}', self.data)

    def chunk(self: Self, chunk_size) -> Iterator[Self]:
        if len(self) % chunk_size != 0:
            raise ValueError(
                f'Cannot chunk data exactly into {chunk_size}: length {len(self)}',
            )
        yield from (
            self[chunk_index * chunk_size:(chunk_index * chunk_size) + chunk_size]
            for chunk_index in range(len(self)//chunk_size)
        )

    def __len__(self: Self) -> int:
        return len(self.data)

    def __getitem__(self: Self, key: int | slice) -> Self:
        return type(self)(
            data=self.data[key],
            endianness=self.endianness,
        )


@dataclasses.dataclass
class Tag:
    code: int
    data_type: int
    count: int
    value: Any

    @classmethod
    def from_bytes(
        cls: type[Self],
        tiff: TIFF,
        tag_bytes: TIFFBytes,
    ) -> Self:
        code, data_type, count = tag_bytes[:8].unpack('HHI')
        return cls(
            code=code,
            data_type=data_type,
            count=count,
            value=cls.unpack_tag_value(tiff, data_type, count, tag_bytes[8:]),
        )

    @staticmethod
    def unpack_tag_value(tiff: TIFF, data_type: int, count: int, value: TIFFBytes) -> Any:
        struct_dtype = DATA_TYPES[data_type]
        size = count * struct.calcsize(struct_dtype)

        if size > len(value):
            offset = value.unpack('I')[0]
            value = tiff.read_bytes(offset, offset+size)
        
        unpacked = value[:size].unpack(struct_dtype * count)

        # if data_type == 2 (ASCII) we want to join the chars together
        if data_type == 2:
            return b''.join(unpacked)
        elif count == 1:
            return unpacked[0]
        return unpacked


class Tags(dict[int, Tag]):
    @classmethod
    def from_tags(cls: type[Self], tags: list[Tag]) -> Self:
        return cls((t.code, t) for t in tags)

    @classmethod
    def from_tiff_bytes(cls: type[Self], tiff: TIFF, tags_bytes: TIFFBytes) -> Self:
        return cls.from_tags([Tag.from_bytes(tiff, tag_bytes) for tag_bytes in tags_bytes.chunk(12)])


@dataclasses.dataclass
class IFD:
    offset: int
    tags: Tags
    next_offset: int

    @classmethod
    def from_tiff_offset(cls: type[Self], tiff: TIFF, offset: int) -> Self:
        tags_start = offset + 2
        tags_count = tiff.read_bytes(offset, tags_start).unpack('H')[0]
        tags_end = tags_start + (tags_count * tag_size)
        tags_bytes = tiff.read_bytes(tags_start, tags_end)
        next_offset = tiff.read_bytes(tags_end, tags_end + 4).unpack('I')[0]
        
        return cls(
            offset=offset,
            tags=Tags.from_tiff_bytes(tiff, tags_bytes),
            next_offset=next_offset,
        )


@dataclasses.dataclass
class TIFF:
    '''Class to help parse TIFF IFDs. Only supports standard TIFFs, not BigTIFF.'''
    href: str
    endianness: Endianness
    ifds: list[IFD]
    
    def __init__(self: Self, href: str) -> None:
        self.href = href

        # we don't use self.read_bytes yet because we don't have endianness
        __bytes = url_read_bytes(self.href, 0, 8)
        self.endianness = Endianness(__bytes[0:2])

        _bytes = TIFFBytes(data=__bytes, endianness=self.endianness)
        
        magic_number = _bytes[2:4].unpack('H')[0]
        if magic_number != 42:
            raise TypeError(f"Unsupported file type: magic number {magic_number} != 42")
        
        self.ifds: list[IFD] = []
        ifd_offset = _bytes[4:8].unpack('I')[0]
        while ifd_offset:
            ifd = self.parse_ifd(ifd_offset)
            self.ifds.append(ifd)
            ifd_offset = ifd.next_offset

    def read_bytes(self: Self, start: int, end: int) -> TIFFBytes:
        # Note that reading for each byte range we want is terribly inefficient.
        # We could instead use some sort of filelike object that will read and cache
        # larger chunks of the file, as needed to accommodate requested byte ranges.
        return TIFFBytes(
            data=url_read_bytes(self.href, start, end),
            endianness=self.endianness,
        )

    def parse_ifd(self: Self, offset: int) -> IFD:
        return IFD.from_tiff_offset(self, offset)

In [396]:
tiff = TIFF(href)
tiff

TIFF(href='https://e84-earth-search-sentinel-data.s3.us-west-2.amazonaws.com/sentinel-2-c1-l2a/10/T/FR/2023/12/S2B_T10TFR_20231223T190950_L2A/B04.tif', endianness=<Endianness.LITTLE_ENDIAN: b'II'>, ifds=[IFD(offset=192, tags={256: Tag(code=256, data_type=3, count=1, value=10980), 257: Tag(code=257, data_type=3, count=1, value=10980), 258: Tag(code=258, data_type=3, count=1, value=16), 259: Tag(code=259, data_type=3, count=1, value=8), 262: Tag(code=262, data_type=3, count=1, value=1), 277: Tag(code=277, data_type=3, count=1, value=1), 284: Tag(code=284, data_type=3, count=1, value=1), 317: Tag(code=317, data_type=3, count=1, value=2), 322: Tag(code=322, data_type=3, count=1, value=1024), 323: Tag(code=323, data_type=3, count=1, value=1024), 324: Tag(code=324, data_type=4, count=121, value=(55962680, 57411167, 58810332, 60222446, 61651003, 63054996, 64463518, 66025043, 67523672, 68987825, 70439668, 71480485, 72831139, 74191906, 75556803, 76922917, 78346396, 79767466, 81177106, 82626646,

## Reading a tile from the image

Read the tile intersecting our POI

In [ ]:
# bits per pixel value
struct.unpack('<H', tag['value'][0:2])[0]

In [ ]:
# compression (259; 1 is uncompressed)
tag = tag_dicts[3]
struct.unpack('<H', tag['value'][0:2])[0]

In [ ]:
# colorspace (1 is b/w)
tag = tag_dicts[4]
struct.unpack('<H', tag['value'][0:2])[0]

In [129]:
# samples per pixel
tag = tag_dicts[5]
struct.unpack('<H', tag['value'][0:2])[0]

1

In [ ]:
# planar configuration (?)
tag = tag_dicts[6]
struct.unpack('<H', tag['value'][0:2])[0]

In [ ]:
# predictor (317)
tag = tag_dicts[7]
struct.unpack('<H', tag['value'][0:2])[0]

In [ ]:
# tile size width (322)
tag = tag_dicts[8]
struct.unpack('<H', tag['value'][0:2])[0]

In [ ]:
# tile size height (323)
tag = tag_dicts[9]
struct.unpack('<H', tag['value'][0:2])[0]

In [ ]:
# tile offsets (324)
tag = tag_dicts[10]
struct_dtype = DATA_TYPES[tag['dtype']]
size = tag['count'] * struct.calcsize(struct_dtype)
offset = struct.unpack('<I', tag['value'])[0]
values = url_read_bytes(href, offset, offset+size)
tile_offsets = struct.unpack('<' + (struct_dtype * tag['count']), values)
print(tile_offsets[:10])

In [ ]:
# tile byte sizes (325)
tag = tag_dicts[11]
struct_dtype = DATA_TYPES[tag['dtype']]
size = tag['count'] * struct.calcsize(struct_dtype)
offset = struct.unpack('<I', tag['value'])[0]
values = url_read_bytes(href, offset, offset+size)
tile_bytes = struct.unpack('<' + (struct_dtype * tag['count']), values)
print(tile_bytes[:10])

In [ ]:
# sample format (339)
tag = tag_dicts[12]
struct.unpack('<H', tag['value'][0:2])[0]

In [ ]:
# GeoTIFF coordinate transformations: http://geotiff.maptools.org/spec/geotiff2.6.html
# pixel scale tag (33550)
tag = tag_dicts[13]
struct_dtype = DATA_TYPES[tag['dtype']]
size = tag['count'] * struct.calcsize(struct_dtype)
offset = struct.unpack('<I', tag['value'])[0]
values = url_read_bytes(href, offset, offset+size)
pixel_scale = struct.unpack('<' + (struct_dtype * tag['count']), values)
print(pixel_scale)

In [ ]:
# tie point tag (33922)
tag = tag_dicts[14]
struct_dtype = DATA_TYPES[tag['dtype']]
size = tag['count'] * struct.calcsize(struct_dtype)
offset = struct.unpack('<I', tag['value'])[0]
values = url_read_bytes(href, offset, offset+size)
tie_point = struct.unpack('<' + (struct_dtype * tag['count']), values)
print(tie_point)

In [ ]:
# GeoTIFF tag structure: http://geotiff.maptools.org/spec/geotiff2.4.html
# geo key directory (34735)
tag = tag_dicts[15]
struct_dtype = DATA_TYPES[tag['dtype']]
size = tag['count'] * struct.calcsize(struct_dtype)
offset = struct.unpack('<I', tag['value'])[0]
values = url_read_bytes(href, offset, offset+size)
geokey_dir = struct.unpack('<' + (struct_dtype * tag['count']), values)
print(geokey_dir)

In [ ]:
## geo double params (34736)
#tag = tag_dicts[16]
#struct_dtype = DATA_TYPES[tag['dtype']]
#size = tag['count'] * struct.calcsize(struct_dtype)
#offset = struct.unpack('<I', tag['value'])[0]
#values = url_read_bytes(href, offset, offset+size)
#geo_doubles = struct.unpack('<' + (struct_dtype * tag['count']), values)
#print(geo_doubles)

In [ ]:
# geo ascii params (34737)
tag = tag_dicts[16]
struct_dtype = DATA_TYPES[tag['dtype']]
size = tag['count'] * struct.calcsize(struct_dtype)
offset = struct.unpack('<I', tag['value'])[0]
values = url_read_bytes(href, offset, offset+size)
geo_asciis = struct.unpack('<' + (struct_dtype * tag['count']), values)
print(geo_asciis)

In [ ]:
# gdal_metadata (42112)
tag = tag_dicts[17]
struct_dtype = DATA_TYPES[tag['dtype']]
size = tag['count'] * struct.calcsize(struct_dtype)
offset = struct.unpack('<I', tag['value'])[0]
values = url_read_bytes(href, offset, offset+size)
gdal_metadata = ''.join([c.decode() for c in struct.unpack('<' + (struct_dtype * tag['count']), values)])
print(gdal_metadata)

In [ ]:
# nodata value (42113)
tag = tag_dicts[18]
struct.unpack('<ss', tag['value'][:2])[0]

In [ ]:
url_read_bytes(href, 4, 1000)

In [ ]:
# TODO: Leave some examples here, but make sure it can't run unless they do this right.
# Show how to extract a tile.
# Show how to find the tile with the POI.
# Show how to display a tile. Switch to leafmap. Write data to a cog. Load into the map.

In [ ]:
block1 = url_read_bytes(href, tile_offsets[0], tile_offsets[0]+tile_bytes[0])

In [ ]:
# need to extract
import zlib
block1_extracted = zlib.decompress(block1)

In [ ]:
block1_array = np.array([struct.unpack('<H', block1_extracted[idx*2:idx*2+2]) for idx in range(len(block1_extracted)//2)]).reshape(1024, 1024)

In [ ]:
block1_array

In [167]:
ifd_offset

8

In [168]:
tag_count

19

In [169]:
tag_size

12

In [182]:
end_of_tags = ifd_offset + 2 + (tag_count * tag_size)

In [203]:
next_ifd_bytes = url_read_bytes(href, end_of_tags, end_of_tags + 4)
next_ifd_bytes

b'\x8d\xa7\xe2_'

In [200]:
next_ifd = struct.unpack('<I', next_ifd_bytes)[0]
next_ifd

0

In [201]:
next_tag_count = struct.unpack('<H', url_read_bytes(href, next_ifd, next_ifd + 2))[0]
next_tag_count

18761

In [202]:
end_of_tags = next_ifd + 2 + (next_tag_count * tag_size)

## Additional exercises to complete on your own time

* Find how many overviews are in this file.
* Find the dimensions and gsd of each overview.
* Repeat reading the tile containing your point of interest, but do so from one of the overviews.
* How can we make reading the file more efficient? Can we get all the IFDs in the file with a single read without having to read in image data?
* Repeat these exercises with a multiband TIFF to see how the file structure differs to support the additional bands.

Any other cool ideas? Let me know and/or share with the group.